cmd/publicize/main.go

```diff
@@ -0,0 +1,188 @@
+		logrus.WithError(err).Fatalf("cannot parse args: '%s'", os.Args[1:])
+	}
+	return o
+}
+
+func (o *options) Validate() error {
+	if (o.gitEmail == "") != (o.gitName == "") {
+		return errors.New("--git-name and --git-email must be specified")
+	}
+
+	for _, group := range []flagutil.OptionGroup{&o.github} {
+		if err := group.Validate(o.dryRun); err != nil {
+			return err
+		}
+	}
+
+	bytes, err := ioutil.ReadFile(o.configPath)
+	if err != nil {
+		return fmt.Errorf("Couldn't read publicize configuration file: %v", o.configPath)
+	}
+
+	var c Config
+	if err := yaml.Unmarshal(bytes, &c); err != nil {
+		return fmt.Errorf("Couldn't unmarshal publicize configuration: %w", err)
+	}
+
+	if err := c.validate(); err != nil {
+		return err
+	}
+
+	o.bonds = make(map[string]string)
+	for _, bond := range c.Bonds {
+		o.bonds[bond.From] = bond.To
+	}
+
+	if err := o.githubEventServerOptions.DefaultAndValidate(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func main() {
+	o := gatherOptions()
+	if err := o.Validate(); err != nil {
+		logrus.Fatalf("Invalid options: %v", err)
+	}
+
+	secretAgent := &secret.Agent{}
+	if err := secretAgent.Start([]string{o.github.TokenPath, o.webhookSecretFile}); err != nil {
+		logrus.WithError(err).Fatal("Error starting secrets agent.")
+	}
+
+	webhookTokenGenerator := secretAgent.GetTokenGenerator(o.webhookSecretFile)
+	githubTokenGenerator := secretAgent.GetTokenGenerator(o.github.TokenPath)
+	getter := func() sets.String {
+		return sets.NewString([]string{string(webhookTokenGenerator()), string(githubTokenGenerator())}...)
+	}
+	logrus.SetFormatter(logrusutil.NewCensoringFormatter(logrus.StandardLogger().Formatter, getter))
+
+	githubClient, err := o.github.GitHubClient(secretAgent, o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting GitHub client.")
+	}
+
+	censor := func(content []byte) []byte {
+		return githubTokenGenerator()
+	}
+	gitClient, err := o.git.GitClient(githubClient, githubTokenGenerator, censor, o.dryRun)
+	if err != nil {
+		logrus.WithError(err).Fatal("Error getting Git client.")
+	}
+
+	logger := logrus.WithField("component", "publicize")
+	serv := &server{
+		githubTokenGenerator: githubTokenGenerator,
+		bonds:                o.bonds,
+		ghc:                  githubClient,
+		gc:                   gitClient,
+		gitName:              o.gitName,
+		gitEmail:             o.gitEmail,
+		remoteHost:           fmt.Sprintf("https://%s", github.DefaultHost),
+		dry:                  o.dryRun,
+	}
+
+	o.githubEventServerOptions.HmacTokenGenerator = webhookTokenGenerator
+	o.githubEventServerOptions.Metrics = githubeventserver.NewMetrics()
+	o.githubEventServerOptions.Logger = logger
+
+	ges := githubeventserver.NewGitHubEventServer(o.githubEventServerOptions)
+	ges.RegisterHandleIssueCommentEvent(serv.handleIssueComment)
+	ges.RegisterHelpProvider(helpProvider, logger)
+
+	interrupts.OnInterrupt(func() {
+		ges.GracefulShutdown()
```

Comment:
Why doesn't this just take a context? `OnInterrupt` is not how this is supposed to work. You want to `interrupts.Run()` and pass in the `ges.Listen()`, allow `Listen()` to shut down when it gets a cancelled context.

Reply:
This is already handled here: https://github.com/kubernetes/test-infra/blob/82eb19f3809bf7b6343291d43ef23005c25edb98/prow/githubeventserver/githubeventserver.go#L96-L97

So handling the cleanup in `interrupts.OnInterrupt` should be sufficient here, right?

Reply:
No. Please write the github event server to run with a context.

Reply:
I will check how `http` handles the context and make the changes.

Reply:
https://github.com/kubernetes/test-infra/pull/18428